lib/ex3/index.ts

```diff
@@ -0,0 +1,221 @@
+  console.log("packet version ", version);
+
+  // Read the ephemeral point, in this example we reuse the same point
+  // which would allow correlation of the onion.
+  const ephemeralPoint = packetReader.readBytes(33);
+  console.log("packet ep      ", ephemeralPoint.toString("hex"));
+
+  // Read the payload from the packet
+  const payload = packetReader.readBytes(packet.length - 33 - 32 - 1);
+  console.log("packet payload ", payload.toString("hex"));
+
+  // Read the HMAC which is the final 32 bytes
+  const hmac = packetReader.readBytes();
+  console.log("packet hmac    ", hmac.toString("hex"));
+
+  // Construct a shared secret using ECDH using the ephemeral point and
+  // our public key.
+  const nodeKey = nodeKeys.shift();
+  const sharedSecret = crypto2.ecdh(ephemeralPoint, nodeKey);
+
+  // Compute an HMAC using the shared secret and the packet (less the
+  // HMAC). We can't compute the HMAC using the full packet, because
+  // the HMAC can't include itself haha.
+  const calcedHmac = crypto2.hmac(
+    sharedSecret,
+    packet.slice(0, packet.length - 32)
+  );
+  console.log("processed hmac ", calcedHmac.toString("hex"));
+
+  // Fail the packet if the HMAC is not the expected value!
+  if (!crypto1.timingSafeEqual(hmac, calcedHmac)) {
+    throw new Error("HMAC failed");
+  }
+
+  // Decrypt the payload using shared secret 
+  // const decrypt_payload = crypto2.ccpDecrypt(
+  //   sharedSecret,
+  //   Buffer.alloc(12, 0x00),          // this nonce value can be randomised and stored in an array to access it later.   
+  //   Buffer.alloc(1, 0x00),          // not sure what to pass in associated data?
+  //   payload
+  // );
+  const decrypt_payload = crypto.chachaDecrypt(sharedSecret,Buffer.alloc(16),payload);
+  // Parse the payload
+  const payloadReader = new BufferReader(decrypt_payload);
+  const hopDataLen = payloadReader.readBigSize();
+  const hopData = payloadReader.readBytes(Number(hopDataLen));
+  const nextHmac = payloadReader.readBytes(32);
+  const nextPayload = payloadReader.eof
+    ? Buffer.alloc(0)
+    : payloadReader.readBytes();
+
+  console.log("payload data     ", hopData.toString("hex"));
+  console.log("payload next hmac", nextHmac.toString("hex"));
+  console.log("payload next data", nextPayload.toString("hex"));
+
+  console.log("");
+
+  // Abort when we don't have any more data
+  if (!nextPayload.length) {
+    return;
+  }
+
+  // Otherwise, return the next packet
+  const packetWriter = new BufferWriter();
+  packetWriter.writeUInt8(0);
+  packetWriter.writeBytes(ephemeralPoint);
+  packetWriter.writeBytes(nextPayload);
+  packetWriter.writeBytes(nextHmac);
+  return packetWriter.toBuffer();
+}
+
+/**
+ * This onion construction adds an HMAC to each layer. The confusing
+ * part is that we process in reverse order. The HMAC for the
+ * inner layer must be supplied to the outer layer. The outer layer
+ * uses the inner layers HMAC as part of the packet construction.
```

Comment:
This comment is really referencing the prior example. While this example builds off the prior example, we should talk about what we do differently here (encrypt the payload).